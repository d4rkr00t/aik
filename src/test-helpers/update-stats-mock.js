/* @flow */

import path from "path";
import fs from "fs";
import { execSync } from "child_process";

export const updateStatsMock = function updateStatsMock(name: string) {
  const fixturesDir = path.join(process.cwd(), "__fixtures__", "webpack-stats", name);
  const pathToFixture = path.join(fixturesDir, "index.js");
  const pathToConfig = path.join(fixturesDir, "webpack.config.js");
  const stats = path.join(fixturesDir, "stats.json");
  try {
    execSync(
      `./node_modules/.bin/webpack ${pathToFixture} --output-path='${fixturesDir}' --mode development --config ${pathToConfig} --json > ${stats}`,
      {
        stdio: "ignore"
      }
    );
  } catch (e) {} // eslint-disable-line
  return JSON.parse(fs.readFileSync(stats, "utf8"));
};
