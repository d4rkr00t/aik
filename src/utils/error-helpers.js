/* @flow */

import chalk from "chalk";

const SYNTAX_ERROR_LABEL = "SyntaxError:";
const SYNTAX_ERROR_LABEL_HUMAN_FRIENDLY = "Syntax Error:";
const ESLINT_PARSE_ERROR = "Parsing error:";

/**
 * Checks whether error is syntax error.
 */
export function isSyntaxError(message: string): boolean {
  return (
    message.indexOf(SYNTAX_ERROR_LABEL) !== -1 ||
    message.indexOf(SYNTAX_ERROR_LABEL_HUMAN_FRIENDLY) !== -1
  );
}

/**
 * Checks whether error is eslint rules warning.
 */
export function isEslintWarning(message: string): boolean {
  return !!message.match(/\d\sproblem/) && !!message.match("warning");
}

/**
 * Checks whether error is eslint parse error.
 */
export function isEslintParseError(message: string): boolean {
  return !!message.match(ESLINT_PARSE_ERROR);
}

export function isDependencyAnExpression(message: string): boolean {
  return !!message.match("the request of a dependency is an expression");
}

export function isPrebuildFileWarning(message: string): boolean {
  return !!message.match("This seems to be a pre-built javascript file");
}

export function findMessagesToFormat(messages: string[]): string[] {
  return messages.filter(message => {
    if (isEslintParseError(message)) {
      return false;
    }

    if (isDependencyAnExpression(message)) {
      return false;
    }

    if (isPrebuildFileWarning(message)) {
      return false;
    }

    return true;
  });
}

export function isFileSnippetLine(line: string): boolean {
  return !!line.match(/\s+\|/);
}

export function isEslintWarningRuleLine(line: string): boolean {
  return !!line.match(/\d+:\d+(.+)warning/);
}

export function getLinePadding(line: string): number {
  return line.split("").findIndex(c => !!c.match(/\S/));
}

export function removeLinePadding(padding: number, line: string): string {
  return line.substring(padding);
}

/**
 * Formats "Syntax Error" message.
 */
export function formatSyntaxError(message: string): string {
  const messageByLine = message.split("\n");
  const padding = getLinePadding(messageByLine[0]);
  const filePath = removeLinePadding(padding, messageByLine[0]);
  const error = removeLinePadding(
    padding,
    messageByLine[1].replace(
      "Module build failed: SyntaxError:",
      SYNTAX_ERROR_LABEL_HUMAN_FRIENDLY
    )
  );
  const snippet = messageByLine
    .filter(isFileSnippetLine)
    .map(removeLinePadding.bind(null, padding));

  return `${chalk.red("Error:")}

${chalk.dim("|")}  ${error}

${chalk.yellow("Where:")} ${filePath}

${snippet.join("\n")}`;
}

/**
 * Formats "ESLint warning" message.
 */
export function formatEslintWarning(message: string): string {
  const messageByLine = message.split("\n");
  const padding = getLinePadding(messageByLine[0]);
  const filePath = removeLinePadding(padding, messageByLine[0]);
  const rule = messageByLine
    .filter(isEslintWarningRuleLine)
    .map(removeLinePadding.bind(null, padding + 2));

  return `${rule.join("\n")}

${chalk.yellow("Where:")} ${filePath}`;
}

/**
 * Beautifies error messages and warnings.
 */
export function formatMessages(messages: string[]): string[] {
  return findMessagesToFormat(messages).map(message => {
    if (isSyntaxError(message)) {
      return formatSyntaxError(message);
    }

    if (isEslintWarning(message)) {
      return formatEslintWarning(message);
    }

    return message;
  });
}
