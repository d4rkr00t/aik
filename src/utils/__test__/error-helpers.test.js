import stripAnsi from "strip-ansi";
import { formatMessages, extractInfoFromModuleNotFoundError } from "../error-helpers";
import { updateStatsMock } from "../../test-helpers/update-stats-mock";

describe("#formatMessages", () => {
  test("Format Syntax Error", () => {
    const stats = updateStatsMock("syntax-error");
    const [error] = formatMessages(stats.errors);
    expect(stripAnsi(error)).toMatchSnapshot();
  });

  test("Format ESLint Warnings", () => {
    const stats = updateStatsMock("eslint-warning");
    const warnings = formatMessages(stats.warnings).join("\n\n\n");
    expect(stripAnsi(warnings)).toMatchSnapshot();
  });

  test("Format Webpack Warnings", () => {
    const stats = updateStatsMock("webpack-warning");
    const warnings = formatMessages(stats.warnings).join("\n\n\n");
    expect(stripAnsi(warnings)).toMatchSnapshot();
  });

  test("Unknown messages", () => {
    const messages = ["some", "custom", "messages"];
    expect(formatMessages(messages)).toMatchSnapshot();
  });
});

describe("#extractInfoFromModuleNotFoundError", () => {
  test("should be able to extract module name and file name", () => {
    const stats = updateStatsMock("module-not-found");
    const { moduleName, file } = extractInfoFromModuleNotFoundError(stats.errors[0]);
    expect(moduleName).toBe("barbarbar");
    expect(file.includes("module-not-found/index.js")).toBe(true);
  });
});
