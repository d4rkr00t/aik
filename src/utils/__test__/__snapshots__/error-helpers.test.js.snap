// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#formatMessages Format ESLint Warnings 1`] = `
"2:3   warning  Unexpected console statement                  no-console
3:27  warning  'item' is assigned a value but never used     no-unused-vars
3:32  warning  Expected to return a value in arrow function  array-callback-return

Where: ./__fixtures__/webpack-stats/eslint-warning/index.js"
`;

exports[`#formatMessages Format Syntax Error 1`] = `
"|  Syntax Error: Unexpected token (2:10)

Where: ./__fixtures__/webpack-stats/syntax-error/index.js

  1 | module.exports = () => {
> 2 |   return <<div>1</div>;
    |           ^
  3 | }
  4 | "
`;

exports[`#formatMessages Format Webpack Warnings 1`] = `""`;

exports[`#formatMessages Unknown messages 1`] = `
Array [
  "some",
  "custom",
  "messages",
]
`;
